Add indexer level task metrics to provide more visibility in the task distribution

rbankar7 opened the pull request:
Adds indexer level task metrics

<!-- Replace XXXX with the id of the issue fixed in this PR. Remove this section if there is no corresponding issue. Don't reference the issue in the title of this pull-request. -->

<!-- If you are a committer, follow the PR action item checklist for committers:
https://github.com/apache/druid/blob/master/dev/committer-instructions.md#pr-and-issue-action-item-checklist-for-committers. -->

### Description

This PR adds indexer level task metrics-
```
"worker/task/running/count"
"worker/task/assigned/count"
"worker/task/completed/count"
```

These metrics will provide more visibility into the tasks distribution across indexers (We often see a task skew issue across indexers and with this issue it would be easier to catch the imbalance)

<!-- Describe the goal of this PR, what problem are you fixing. If there is a corresponding issue (referenced above), it's not necessary to repeat the description here, however, you may choose to keep one summary sentence. -->

<!-- Describe your patch: what did you change in code? How did you fix the problem? -->

<!-- If there are several relatively logically separate changes in this PR, create a mini-section for each of them. For example: -->


This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/WorkerTaskManagerMonitor.java:
```diff
@@ -0,0 +1,81 @@
+import java.util.Map;
+import java.util.function.Function;
+
+public class WorkerTaskManagerMonitor extends AbstractMonitor
```
Can we reuse WorkerTaskCountStatsMonitor itself?

rbankar7 replied:
Wanted to emit this metric from indexer pods
We have seen frequent missing metrics issues when we enable more metrics which emit from the overlord

abhishekagarwal87 replied:
Are those stats not available in indexer? You could selectively emit stats that are available on indexers but use the same monitor.

kfaraz replied:
+1 on using the existing `WorkerTaskCountStatsMonitor` as the new metrics fall in the same bucket.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/WorkerTaskManagerMonitor.java:
```diff
@@ -0,0 +1,81 @@
+  @Override
+  public boolean doMonitor(ServiceEmitter emitter)
+  {
+    final Map<String, Integer> runningTasks, assignedTasks, completedTasks;
+
+    runningTasks = getDataSourceTasks(workerTaskManager.getRunningTasks(), WorkerTaskManager.TaskDetails::getDataSource);
+    assignedTasks = getDataSourceTasks(workerTaskManager.getAssignedTasks(), Task::getDataSource);
+    completedTasks = getDataSourceTasks(workerTaskManager.getCompletedTasks(), TaskAnnouncement::getTaskDataSource);
```
Preferred style in Druid (if not Java itself):
```suggestion
    final Map<String, Integer> runningTasks = getDataSourceTasks(workerTaskManager.getRunningTasks(), WorkerTaskManager.TaskDetails::getDataSource);
    final Map<String, Integer> assignedTasks = getDataSourceTasks(workerTaskManager.getAssignedTasks(), Task::getDataSource);
    final Map<String, Integer> completedTasks = getDataSourceTasks(workerTaskManager.getCompletedTasks(), TaskAnnouncement::getTaskDataSource);
```

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/WorkerTaskManager.java:
```diff
@@ -179,6 +179,16 @@ public Map<String, TaskAnnouncement> getCompletedTasks()
 
+  public Map<String, TaskDetails> getRunningTasks()
+  {
+    return runningTasks;
```
No need to expose the `TaskDetails` class as no other field of the `TaskDetails` class is being used in the monitoring. Might as well just do the following :

```suggestion
    return CollectionUtils.mapValues(runningTasks, detail -> detail.task);
```

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/WorkerTaskManagerMonitor.java:
```diff
@@ -0,0 +1,81 @@
+    }
+  }
+
+  private <T> Map<String, Integer> getDataSourceTasks(Map<String, T> taskMap, Function<T, String> getDataSourceFunc)
```
A better name would be `getDatasourceToNumTasks` or `getNumTasksPerDatasource`.

Also, the first argument to this method should be only the values of the map rather than the whole map. The key of the map (taskId) is never used in this method.

kfaraz commented on indexing-service/src/test/java/org/apache/druid/indexing/worker/WorkerTaskManagerMonitorTest.java:
```diff
@@ -0,0 +1,101 @@
+  @Before
+  public void setUp()
+  {
+    task = createMock(Task.class);
```
Use `NoopTask` instead of creating a mock.

kfaraz commented on indexing-service/src/test/java/org/apache/druid/indexing/worker/WorkerTaskManagerMonitorTest.java:
```diff
@@ -0,0 +1,101 @@
+    taskDetails = createMock(WorkerTaskManager.TaskDetails.class);
+    EasyMock.expect(taskDetails.getDataSource()).andReturn("dummy_DS2");
+    EasyMock.replay(taskDetails);
+    taskAnnouncement = createMock(TaskAnnouncement.class);
```
Try to use concrete instances of `TaskAnnouncement` and `TaskDetails` (if needed) instead of mocks.

kfaraz commented on indexing-service/src/test/java/org/apache/druid/indexing/worker/WorkerTaskManagerMonitorTest.java:
```diff
@@ -0,0 +1,101 @@
+    EasyMock.replay(taskAnnouncement);
+    taskRunner = createMock(TaskRunner.class);
+    taskConfig = createMock(TaskConfig.class);
+    overlordClient = createMock(OverlordClient.class);
```
Use `NoopOverlordClient` instead.

kfaraz reviewed (commented):
Thanks for your first PR, @rbankar7 ! I have left some suggestions.

kfaraz reviewed (commented):
Thanks for the update, @rbankar7 . 
I have left some comments that need to be addressed.
Please update the docs to add info about the new metrics.

rbankar7 commented on server/src/test/java/org/apache/druid/server/metrics/WorkerTaskCountStatsMonitorTest.java:
```diff
@@ -193,6 +233,45 @@ public void testMonitor()
+    Assert.assertEquals(6, emitter.getEvents().size());
+    emitter.verifyValue(
+        "worker/task/running/count",
+        ImmutableMap.of("dataSource", "dummyDS1"),
```
Ok will change it 
Thanks

rbankar7 commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/WorkerTaskManager.java:
```diff
@@ -179,6 +186,16 @@ public Map<String, TaskAnnouncement> getCompletedTasks()
+  public Map<String, Task> getAssignedTasks()
+  {
+    return assignedTasks;
+  }
```
removed these methods as these were getting called only once by `getWorkerRunningTasks` and `getWorkerAssignedTasks`

rbankar7 commented:
> Thanks for the update, @rbankar7 . I have left some comments that need to be addressed. Please update the docs to add info about the new metrics.

Thanks @kfaraz for the review 
I have addressed the comments and updated the doc with the new metrics

kfaraz commented on server/src/main/java/org/apache/druid/server/metrics/IndexerTaskCountStatsProvider.java:
```diff
@@ -0,0 +1,44 @@
+public interface IndexerTaskCountStatsProvider
+{
+  /**
+   * The number of running tasks on an indexer
```
Minor suggestion for the javadocs but this can be addressed later.
```suggestion
   * Map from datasource name to the number of running tasks on the Indexer.
```

kfaraz reviewed (approved):
Thanks for adding the new metric and docs, @rbankar7 !

kfaraz commented:
@rbankar7 , the checkstyle and `WorkerTaskCountStatsMonitorTest` are failing. Please take a look.

kfaraz commented:
@rbankar7 , there are more checkstyle failures. To ensure that you catch all of them, please run the following on your local machine before you commit:

```
mvn checkstyle:checkstyle
```

rbankar7 commented:
@kfaraz I fixed the issues and verified that there are no checkstyle errors
Thanks!

kfaraz commented:
@rbankar7 , I ran the workflow. The test`WorkerTaskCountStatsMonitorTest.testMonitorNotMiddleManager` seems to be failing.

```java
Error:  Errors: 
Error:  org.apache.druid.server.metrics.WorkerTaskCountStatsMonitorTest.testMonitorNotMiddleManager
Error:    Run 1: WorkerTaskCountStatsMonitorTest.testMonitorNotMiddleManager:288 » Configuration Guice configuration errors:

1) No implementation for org.apache.druid.server.metrics.IndexerTaskCountStatsProvider was bound.
  while locating org.apache.druid.server.metrics.IndexerTaskCountStatsProvider
```

The code coverage also seems insufficient for `WorkerTaskManager`.
<img width="905" alt="Screenshot 2024-03-15 at 4 55 00 PM" src="https://github.com/apache/druid/assets/18635897/50a3d709-f938-41ff-952d-c66a42bcd19b">

rbankar7 commented:
> I ran the workflow. The testWorkerTaskCountStatsMonitorTest.testMonitorNotMiddleManager seems to be failing.
@kfaraz will fix it
Thanks

rbankar7 commented:
@kfaraz looks like build is passing now
could you please take a look 
Thanks!

kfaraz commented:
Thanks a lot for the fix, @rbankar7 !
Congrats on your first commit to Apache Druid! 🎉
